loopy/schedule/checker/schedule.py

```diff
@@ -214,36 +341,411 @@ def generate_pairwise_schedules(
+    lid_lex_dim_names = set()
+    gid_lex_dim_names = set()
+    par_iname_constraint_dicts = {}
+    for iname in knl.all_inames():
+        ltag = knl.iname_tags_of_type(iname, LocalIndexTag)
+        if ltag:
+            assert len(ltag) == 1  # (should always be true)
+            ltag_var = LTAG_VAR_NAMES[ltag.pop().axis]
+            lid_lex_dim_names.add(ltag_var)
+            par_iname_constraint_dicts[iname] = {1: 0, iname: 1, ltag_var: -1}
+
+            continue  # Shouldn't be any GroupIndexTags
+
+        gtag = knl.iname_tags_of_type(iname, GroupIndexTag)
+        if gtag:
+            assert len(gtag) == 1  # (should always be true)
+            gtag_var = GTAG_VAR_NAMES[gtag.pop().axis]
+            gid_lex_dim_names.add(gtag_var)
+            par_iname_constraint_dicts[iname] = {1: 0, iname: 1, gtag_var: -1}
+
+    # Sort for consistent dimension ordering
+    lid_lex_dim_names = sorted(lid_lex_dim_names)
+    gid_lex_dim_names = sorted(gid_lex_dim_names)
+
+    # }}}
+
+    # {{{ Intra-group and global blex ("barrier-lex") order creation
+
+    # (may be combined with pass above in future)
+
+    """In blex space, we order barrier-delimited sections of code.
+    Each statement instance within a single barrier-delimited section will
+    map to the same blex point. The resulting statement instance ordering
+    (SIO) will map each statement to all statements that occur in a later
+    barrier-delimited section.
+
+    To achieve this, we will first create a map from statement instances to
+    lexicographic space almost as we did above in the intra-thread case,
+    though we will not increment the fastest-updating lex dim with each
+    statement, and we will increment it with each barrier encountered. To
+    denote these differences, we refer to this space as 'blex' space.
+
+    The resulting pairwise schedule, if composed with a map defining a
+    standard lexicographic ordering to create an SIO, would include a number
+    of unwanted 'before->after' pairs of statement instances, so before
+    creating the SIO, we will subtract unwanted pairs from a standard
+    lex order map, yielding the 'blex' order map.
+    """
 
-    from loopy.schedule.checker.utils import (
-        sorted_union_of_names_in_isl_sets,
-        create_symbolic_map_from_tuples,
-        add_dims_to_isl_set,
-    )
+    # {{{ Get upper and lower bound for each loop that contains a barrier
+
+    iname_bounds_pwaff = {}
+    for iname in loops_with_barriers["local"] | loops_with_barriers["global"]:
+        bounds = knl.get_iname_bounds(iname)
+        iname_bounds_pwaff[iname] = (
+            bounds.lower_bound_pw_aff, bounds.upper_bound_pw_aff)
+
+    # }}}
+
+    # {{{ Create blex order maps and blex tuples defining statement ordering (x2)
+
+    all_par_lex_dim_names = lid_lex_dim_names + gid_lex_dim_names
+
+    # {{{ _gather_blex_ordering_info(sync_kind): gather blex info for sync_kind
+
+    def _gather_blex_ordering_info(sync_kind):
```

Comment:
Could you promote this function to the top level? I find nested hard to read for functions this long. (Specifically, it's hard to see what data goes in.)

Reply:
Ok ([diff here](https://github.com/inducer/loopy/compare/0e8be5b..6fab4bc)).